lib/config.js

```diff
@@ -411,34 +417,59 @@ function parseConfig (configFilePath, cliOptions, parseOptions) {
   // add the user's configuration in
   config.set(cliOptions)
 
+  let configModuleReturn
   try {
-    configModule(config)
+    configModuleReturn = configModule(config)
   } catch (e) {
     return fail('Error in config file!\n', e)
   }
+  function finalizeConfig (config) {
+    // merge the config from config file and cliOptions (precedence)
+    config.set(cliOptions)
 
-  // merge the config from config file and cliOptions (precedence)
-  config.set(cliOptions)
-
-  // if the user changed listenAddress, but didn't set a hostname, warn them
-  if (config.hostname === null && config.listenAddress !== null) {
-    log.warn(`ListenAddress was set to ${config.listenAddress} but hostname was left as the default: ` +
+    // if the user changed listenAddress, but didn't set a hostname, warn them
+    if (config.hostname === null && config.listenAddress !== null) {
+      log.warn(`ListenAddress was set to ${config.listenAddress} but hostname was left as the default: ` +
       `${defaultHostname}. If your browsers fail to connect, consider changing the hostname option.`)
-  }
-  // restore values that weren't overwritten by the user
-  if (config.hostname === null) {
-    config.hostname = defaultHostname
-  }
-  if (config.listenAddress === null) {
-    config.listenAddress = defaultListenAddress
-  }
+    }
+    // restore values that weren't overwritten by the user
+    if (config.hostname === null) {
+      config.hostname = defaultHostname
+    }
+    if (config.listenAddress === null) {
+      config.listenAddress = defaultListenAddress
+    }
 
-  // configure the logger as soon as we can
-  logger.setup(config.logLevel, config.colors, config.loggers)
+    // configure the logger as soon as we can
+    logger.setup(config.logLevel, config.colors, config.loggers)
 
-  log.debug(configFilePath ? `Loading config ${configFilePath}` : 'No config file specified.')
+    log.debug(configFilePath ? `Loading config ${configFilePath}` : 'No config file specified.')
 
-  return normalizeConfig(config, configFilePath)
+    return normalizeConfig(config, configFilePath)
+  }
+  const returnIsThenable = configModuleReturn != null && typeof configModuleReturn === 'object' && typeof configModuleReturn.then === 'function'
+  if (returnIsThenable) {
+    if (promiseConfig !== true) {
+      const errorMessage =
+        'The `parseOptions.promiseConfig` option must be set to `true` to ' +
+        'enable promise return values from configuration files. ' +
+        'Example: `parseConfig(path, cliOptions, { promiseConfig: true })`'
+      return fail(errorMessage)
+    }
+    return configModuleReturn.then(function onKarmaConfigModuleFulfilled (/* ignoredResolutionValue */) {
```

Comment:
```suggestion
    return Promise.resolve(configModuleReturn).then(function onKarmaConfigModuleFulfilled (/* ignoredResolutionValue */) {
```

as otherwise, it crashes when `promiseConfig`, but config is not async.

Reply:
This should be handled by: 
https://github.com/karma-runner/karma/blob/5d92865dd3d61c9570bd85102e56352d9dc0715c/lib/config.js#L450-L451

and then by:
https://github.com/karma-runner/karma/blob/5d92865dd3d61c9570bd85102e56352d9dc0715c/lib/config.js#L462-L472

Though I may have made this more complicated than it needed to be (as I just mentioned in a different comment). In hindsight, I think there may be a more clear way.

```diff
-  const returnIsThenable = configModuleReturn != null && typeof configModuleReturn === 'object' && typeof configModuleReturn.then === 'function'
-  if (returnIsThenable) {
-    if (promiseConfig !== true) {
-      const errorMessage =
-        'The `parseOptions.promiseConfig` option must be set to `true` to ' +
-        'enable promise return values from configuration files. ' +
-        'Example: `parseConfig(path, cliOptions, { promiseConfig: true })`'
-      return fail(errorMessage)
-    }
-    return configModuleReturn.then(function onKarmaConfigModuleFulfilled (/* ignoredResolutionValue */) {
-      return finalizeConfig(config)
-    })
-  } else {
-    if (promiseConfig) {
+  if (promiseConfig) {
+    const returnIsThenable = configModuleReturn != null && typeof configModuleReturn === 'object' && typeof configModuleReturn.then === 'function'
+    if (returnIsThenable) {
+      return configModuleReturn.then(function onKarmaConfigModuleFulfilled (/* ignoredResolutionValue */) {
+        return finalizeConfig(config)
+      })
+    } else {
       try {
         return Promise.resolve(finalizeConfig(config))
       } catch (exception) {
         return Promise.reject(exception)
       }
     } else {
       return finalizeConfig(config)
     }
   }
 }
```

Reply:
-   the config file function _is not required to return a promise_ in order for `parseConfig` to return a promise.
-   if a promise is not returned by the function, the config is handled the same way as synchronous parsing and an immediately resolved promise is returned.

A promise resolving just tells us that the asynchronous code is done. We don't use the return/fulfillment values otherwise since we already have the config object and expect it to be modified via `config.set`.

---

I guess this raises another question. Assumptions I made in the first and second passes at the code may not be in line with what you were expecting.

Do we want to require the config file to have a promise as a return value?

Reply:
Re original suggestion, I guess I did something wrong. I've retested now and everything seems to work as intended. So never mind. Thanks for the explanation.

> Though I may have made this more complicated than it needed to be (as I just mentioned in a different comment). In hindsight, I think there may be a more clear way.

No, I think the current code is good and it helps the user to resolve the problem.

> Do we want to require the config file to have a promise as a return value?

No, we don't want to require it to return a promise. The synchronous config should work with the `promiseConfig == true`.

I'll resolve this thread as there is nothing to do here and respond in another one in a minute.